flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java

```diff
@@ -1293,6 +1328,121 @@ private TableResult buildShowResult(String columnName, String[] objects) {
+                                                    watermarkSpec
+                                                            .getWatermarkExpression()
+                                                            .asSummaryString()))
+                            .collect(Collectors.joining("\n")));
+        }
+        // append constraint
+        if (schema.getPrimaryKey().isPresent()) {
+            sb.append(",\n");
+            sb.append(String.format("%s%s", printIndent, schema.getPrimaryKey().get()));
+        }
+        sb.append("\n) ");
+        // append comment
+        String comment = table.getComment();
+        if (StringUtils.isNotEmpty(comment)) {
+            sb.append(String.format("COMMENT '%s'\n", comment));
+        }
+        // append partitions
+        ResolvedCatalogTable catalogTable = (ResolvedCatalogTable) table;
+        if (catalogTable.isPartitioned()) {
+            sb.append("PARTITIONED BY (")
+                    .append(
+                            catalogTable.getPartitionKeys().stream()
+                                    .map(key -> String.format("`%s`", key))
+                                    .collect(Collectors.joining(", ")))
+                    .append(")\n");
+        }
+        // append `with` properties
+        Map<String, String> options = table.getOptions();
+        sb.append("WITH (\n")
+                .append(
+                        options.entrySet().stream()
+                                .map(
+                                        entry ->
+                                                String.format(
+                                                        "%s'%s' = '%s'",
+                                                        printIndent,
+                                                        entry.getKey(),
+                                                        entry.getValue()))
+                                .collect(Collectors.joining(",\n")))
+                .append("\n)\n");
+        return sb.toString();
+    }
+
+    private String getColumnString(Column column) {
+        final StringBuilder sb = new StringBuilder();
+        sb.append(EncodingUtils.escapeIdentifier(column.getName()));
+        sb.append(" ");
+        // skip data type for computed column
+        if (column instanceof Column.ComputedColumn) {
+            sb.append(
+                    column.explainExtras()
+                            .orElseThrow(
+                                    () ->
+                                            new TableException(
+                                                    String.format(
+                                                            "Column expression can not be null for computed column '%s'",
+                                                            column.getName()))));
+        } else {
+            DataType dataType = column.getDataType();
+            String type = dataType.toString();
```

Comment:
Use `dataType#asSerializableString` instead of `toString`/`asSummaryString`. `asSerializableString` doesn't show timestamp kind and we don't need the following special logic then. 

Actually, we should always use `asSerializableString`, rather than `asSummaryString`, because only the serialized string is fully supported to be parsed, see `LogicalTypeParser`. 

Besides, it would be better to avoid using `toString` and explicitly use `asSerializableString`, otherwise, it's hard to trace where the `asSerializableString`/`asSummaryString` are used.

Reply:
> Use `dataType#asSerializableString` instead of `toString`/`asSummaryString`. `asSerializableString` doesn't show timestamp kind and we don't need the following special logic then.
> 
> Actually, we should always use `asSerializableString`, rather than `asSummaryString`, because only the serialized string is fully supported to be parsed, see `LogicalTypeParser`.
> 
> Besides, it would be better to avoid using `toString` and explicitly use `asSerializableString`, otherwise, it's hard to trace where the `asSerializableString`/`asSummaryString` are used.

Hi Jark, I agree with you that `asSerializableString` goes first. Just one thing to confirm, `LogicalTypeParser#parseTypeByKeyword` uses `VarCharType.MAX_LENGTH` as varchar length, and for `VarCharType#asSerializableString` will always print string data type as `VARCHAR(2147483647)`.

Reply:
If this is acceptable, then `column.getDataType().getLogicalType().asSerializableString()` is definitely better.

Reply:
I think we can use `STRING` as serializable string for `VarCharType.MAX_LENGTH` too. What do you think @twalthr ?

Reply:
Another case is for `ts timestamp_ltz(3)`, the summary format is `TIMESTAMP_LTZ(%d)`, and the serializable format is `TIMESTAMP(%d) WITH LOCAL TIME ZONE`

Reply:
Since the case-by-case check is not attractive, I'm just wondering do we need to preserve the original statement made by users as much as possible? If so, maybe we should add another API method to fulfill this requirement.

Reply:
I think we can leave this into another issue. `STRING`, `BYTES`, `TIMESTAMP_LTZ` are just abbreviations, it's fine to just display `VARCHAR(MAX)`, `TIMESTAMP(%d) WITH LOCAL TIME ZONE`, because they are single source of truth.